### ISSUE TITLE
possible to install Ubuntu 14.04?

### ISSUE THREAD

**Issue by mo18** [Microsoft/WSL#2321]
I'm running windows version 1703 (15063.413). I went to re-install my bash environment using `lxrun /uninstall /full` and `lxrun /install`. When I did that it installed 16.04. Is there a way to get 14.04 installed with this version?

**Comment by benhillis** [comment]
@mo18 - no there is not.  Can you describe why you want 14.04 versus 16.04 to help us understand your scenario?

**Comment by therealkenc** [comment]
Ben is right; your scenario matters. For some valid scenarios that matter, you might want [chroot](http://man7.org/linux/man-pages/man1/chroot.1.html), which works alright on WSL these days.

Begs a good question though. Xenial won't end-of-life until 2021, but there's no way everyone is going to tolerate staying on 16.04 that long. Meanwhile others _won't_ want to upgrade, for policy reasons. So in some foreseeable future WSL will have to be prepared for more than one version in the store. You can probably get away with "install 16.04 and then upgrade" for a _while_, but not all the way through 2021.

**Comment by benhillis** [comment]
@therealkenc - that's one of the great things about WSL distros being in the store. Each of our distro vendors are now in control of what images they want to be available to the customers.  You could imagine an Ubuntu store app that on first launch asks you which version you want to install (Trusty, Xenial, etc).

**Comment by therealkenc** [comment]
That is one of the great things about WSL distros. My first thought when I saw Rich's blog post in May was... okay, so when is Windows 10 user space going to work like that? 😜

**Comment by mo18** [comment]
@benhillis - our production stack runs on 14.04, so we want to be able to keep all versions within our stack the same. I suppose it's not a requirement, but it sure helps when things aren't working properly

**Comment by sunjoong** [comment]
@mo18 - https://github.com/RoliSoft/WSL-Distribution-Switcher is an another way to install other distributions. They says it for ubuntu, "ubuntu – 12.04.5, 12.04, precise-20160707, precise | 14.04.5, 14.04, trusty-20160802, trusty | 16.04, xenial-20160809, xenial, latest | 16.10, yakkety-20160806.1, yakkety, devel".

**Comment by fpqc** [comment]
To be honest, you should be able to do it if you know what you're doing.  Just get a hold of the 14.04 docker image, untar it inside of /root as root, and replace the files in the rootfs directly from Windows.  This should still work with the Store-based Ubuntu

**Comment by sunilmut** [comment]
@dustinkirkland might be interested in this conversation.

**Comment by mo18** [comment]
Thanks you all for your responses. I'm definitely curious to see if @dusrinkirkland has a response. 

To expand on my need for this, our developers currently use Mac's and build out their dev environments using homebrew and pip to get all of the dependencies installed. There are a bunch of challenges with versions of software to getting the environment operational. We're investigating docker and other solutions in order to standardize our setup. When I heard about WSL and I started messing around with it, I decided to switch from Mac to Windows. I'm in more of a devops role and having Windows (I do a decent amount with powershell) with an Ubuntu shell integrated on my desktop simplifies my workflow considerably. I started pitching the idea to some team members recently and today I decided to setup the dev environment that the rest of the development team use. This is when I ran into my issue. I had installed a bunch of unnecessary software and wanted a clean slate to start with so I uninstalled and reinstalled the WSL environment. 

I'm not sure that 16.04 is a show stopper for me as I'm sure we'll be moving to that in the near future. It's also probably a step in the right direction when compared to using homebrew. 

Thanks again for the feedback...

**Comment by dustinkirkland** [comment]
Well, the overall experience with Ubuntu 16.04 is going to be *way* better.  More applications, newer tools, updated compilers and libraries, more active bug fixing/updates than 14.04.  Thus, we took the decision to move WSL to the "latest" LTS, after a few months of burn in and scale testing.

For this use case (must have 14.04 with a specific package set), I'd recommend using a VM or Docker.

**Comment by NicolasDorier** [comment]
Reason why I need trusty: https://github.com/bitcoin/bitcoin/issues/10269 

TL;DR: I can't cross compile for windows on Xenial.

**Comment by fpqc** [comment]
@NicolasDorier very simple to fix that problem, and trusty won't fix it: before compiling, edit your PATH to remove the appended windows system paths.  In the latest Insider build, there is a way to disable interop completely. If you're having trouble building something like this, you should file a bug report rather than use a workaround that likely won't work.

**Comment by NicolasDorier** [comment]
@fpqc this does not work in my case. The bug is caused by Xenial's mingw32. I need Trusty to compile correctly.
I am using Docker just because of that.

**Comment by fpqc** [comment]
@NicolasDorier Does this bug happen on Ubuntu natively as well?  If not, could you please submit a bug report here for mingw32?

**Comment by NicolasDorier** [comment]
yes it does happen natively, I am just responding to why offering support for Ubuntu 14.04 is important.

As @benhillis was asking

> @mo18 - no there is not. Can you describe why you want 14.04 versus 16.04 to help us understand your scenario?

This is a WSL issue in so far that using Ubuntu 14.04 should be supported. 
As a user I agree that new distro are better. As a developer, I just want to use whatever is needed to get my code to build, and WSL target developers.

**Comment by fpqc** [comment]
@NicolasDorier If it's an Ubuntu bug, have you submitted the bug to the package maintainer or to Canonical or even upstream?

**Comment by MikeGitb** [comment]
@NicolasDorier: Maybe you linked to the wrong issue? It is about a problem caused by wsl spending windows path to the bash path variable, not about a problem with the mingw installation on Ubuntu 16.04

**Comment by NicolasDorier** [comment]
The issue is about getting making it possible to install Ubuntu 14.04, and I am telling why it is useful.
The reason why my case is failing is not as important as the fundamental reason why WSL need to support 14.04.

As developer, there is such kind of bug daily, and WSL goal is; I think, to help windows developer to bridge with the linux world.
Telling me "it is mingw problem" is missing the point, my point is that it does not matter, WSL should just support ubuntu 14.04 as some build environments just can't update to the latest distros.

**Comment by benhillis** [comment]
@nicolasdorier - We do appreciate the feedback. I took a look at the thread you linked and replied, I think I know what the issue with the bitcoin project and have provided instructions how to fix the issue.  If you have the bitcoin project set up could you try the fix I suggest?

**Comment by benhillis** [comment]
@NicolasDorier - I was able to get around this issue by modifying the funcs.mk file to wrap "$PATH" with quotes "$PATH"

https://github.com/bitcoin/bitcoin/issues/10269#issuecomment-297511480

**Comment by MikeGitb** [comment]
@NicolasDorier: The reason I was asking is that switching back to ubuntu14.04 wouldn't -as far as I can tell - resolve the issue you linked to.

Personally I'd also very much appreciate if I could stay in sync with the Linux machines in my environment, which regrettably still use ubuntu14.04.  So I can absolutely understand your general scenario. Just that particular issue you linked to seems to have nothing to do with the Ubuntu version and it is also not the result of a newly introduced mingw bug in xenial.

**Comment by therealkenc** [comment]
@MikeGitb
> Personally I'd also very much appreciate if I could stay in sync with the Linux machines in my environment, which regrettably still use ubuntu14.04

It's a pretty reasonable ask. Say the team fixes #1878, and the next day I upgrade to Zesty. Someone posts a bug that manifests on Xenial. I should be able to install Xenial side-by-side, maybe install their crazy repro deps, and take a quick look without jumping through hoops.

@benhillis 
> You could imagine an Ubuntu store app that on first launch asks you which version you want to install (Trusty, Xenial, etc).

Yeah but what if I want both. While I was doing the EnumWSL thing I realized "nah, this isn't going to work".

We're still missing a few pieces to do this right.
* `mount -t lxfs` ([message](https://github.com/Microsoft/BashOnWindows/issues/2212#issuecomment-309818878)). Otherwise there's a lot of dupe posts coming to a github near you.
* Stubs for the device surface required to make `systemd` happy. This sounds like a big ask, and it is. But less so if `systemd` thinks there's no devices to fire up. User space just thinks it's running on the Worst Motherboard Ever. The machine is de-facto diskless and `LxFs` is analgous to NFS as far as user space is concerned. No usb, no sound, no graphics. `lsdsk` `lsusb` et al work fine (#2287), but return nothing.
* PID namespaces ([how hard can it be](http://s2.quickmeme.com/img/ca/ca1b7c6db7cfa242127917a9ace57901d243d9fae6c27f77debafcbee595f3b6.jpg)).
* [`pivot_root(2)`](http://man7.org/linux/man-pages/man2/pivot_root.2.html)

I think you know where I am going with this. Anyway, I'll put down the bong now....

**Comment by NicolasDorier** [comment]
@benhillis sorry, I linked you to the wrong issue 
https://github.com/bitcoin/bitcoin/issues/8732

See the error you would have by compiling  with Xenial https://github.com/bitcoin/bitcoin/issues/8511
You can reproduce easily.

**Comment by NicolasDorier** [comment]
@MikeGitb my bad, the git issue I linked was the wrong one see bitcoin/bitcoin#8732 or  bitcoin/bitcoin#8511

**Comment by benhillis** [comment]
@NicolasDorier - Thanks for the link.  Reading over that thread it seems to me there is probably some workaround that can be used to get things working again, perhaps and older version of gcc or some updates to the makefile?

Thanks for the feedback about distribution choice, definitely something for us and our partners (like @dustinkirkland at Canonical) to consider.

**Comment by NicolasDorier** [comment]
Fanquake tried several alternatives in the thread and everything failed. For now compiling Bitcoin on Xenial for Windows is just not supported.

I think the underlying reason matter less than the fact that some use case still need specific old distros.
If it is not a huge architectural change to WSL, it is worth it as a feature, this would be highly useful for developers. I really hope to get rid of docker for my development need.

**Comment by molxyz** [comment]
Hello, 
I'm having the same issue as in https://github.com/bitcoin/bitcoin/issues/8511; however, I'm wondering if 
 Ubuntu 16.04.3 LTS can be improved to do everything that Ubuntu 14.04 Trusty does and more? If so it would be great to keep this version. Thank you.

**Comment by therealkenc** [comment]
> Ubuntu 16.04.3 LTS can be improved to do everything that Ubuntu 14.04 Trusty does and more?

This issue isn't with Ubuntu (read Debian, read "Linux"). Someone just needs to bang on Bitcoin Core. This isn't a WSL thing.

That said, running Ubuntu 14 and 16 side-by-side remains a valid use case. People do this with VMs all the time. Sharing `/home` (or `/var/downloads/GoT`) between the two is also a given requirement as far as I am concerned. I am sure they're looking at options for supporting this. In the mean time you can use RolliSoft's [Distribution Switcher](https://github.com/RoliSoft/WSL-Distribution-Switcher), which is similar to going the Docker/VM route.

**Comment by MikeGitb** [comment]
>  Sharing /home (or /var/downloads/GoT) between the two is also a given requirement as far as I am concerned.
I think that depends a little on what distributions we are talking about. Sure, I'd e.g. like to have the same ssh clinet config between ubuntu 14.04 and 16.04, but accross different kinds of distributions (Ubuntu fedora) there might actually be a lot of conflicting configurations, so there I'd like to only share things like "documents" and "downloads".

**Comment by NicolasDorier** [comment]
```
This issue isn't with Ubuntu (read Debian, read "Linux"). Someone just needs to bang on Bitcoin Core. This isn't a WSL thing.
```

WSL is here to make cross development possible. The reality is that cross development do not always support the latest ubuntu version. This is an issue of Bitcoin Core, yes, but it is also an issue of WSL, as WSL should, as part of its mission, make it possible to build projects coming from different dev environment.

**Comment by thodt-md** [comment]
I use `virt-manager` to connect to Spice VM guests. With 14.04 WSL everything works fine, with 16.04 the screen connection fails with a message described here - https://bugs.launchpad.net/ubuntu/+source/virt-manager/+bug/1443954 (no, the workaround mentioned there did _not_ work for me, in case anyone is wondering).

This may or may not be a bug in `virt-manager`, but at the same time the connection succeeds if I do it on a native 16.04 Ubuntu machine.

I do not try redirecting USB, the VM in question does not even have a redirector device installed.

This is probably a niche use case, but it would still benefit from having a choice to install an older Ubuntu version. Newer software versions tend to use features, that may not be implemented in WSL, so it's not always better to use the latest LTS, IMHO.

**Comment by avilaHash** [comment]
+1 on adding support for choosing an earlier Ubuntu version. Facing the same issue as Nicolas

**Comment by therealkenc** [comment]
Discussion pretty much ran course, if there is such a thing. It is possible to run Ubuntu 14.04 on WSL with [Distribution Switcher](https://github.com/RoliSoft/WSL-Distribution-Switcher), or in a [`chroot`](http://man7.org/linux/man-pages/man1/chroot.1.html) environment, and the like, but this is currently not an officially supported scenario.
